src/skmultiflow/meta/dynamic_weighted_majority.py

```diff
@@ -0,0 +1,271 @@
+            entry is associated with the X entry of the same index.
+        """
+        return self._aggregate_expert_predictions(
+                    self.get_expert_predictions(X))
+
+    def fit_single_sample(self, X, y, classes=None, weight=None):
+        """
+        Fits a single sample of shape X.shape=(1, n_attributes) and y.shape=(1)
+
+        Aggregates all experts' predictions, diminishes weight of experts whose
+        predictions were wrong, and may create or remove experts every _period_
+        samples.
+
+        Finally, trains each individual expert on the provided data.
+
+        Train loop as described by Kolter and Maloof [1]_.
+
+
+        Parameters
+        ----------
+        X: Numpy.ndarray of shape (n_samples, n_features) 
+            Features matrix used for partially updating the model.
+            
+        y: Array-like
+            An array-like of all the class labels for the samples in X.
+            
+        classes: list 
+            List of all existing classes. This is an optional parameter.
+
+        weight: None
+            Instance weight. This is ignored by the ensemble and is only
+            for compliance with the general skmultiflow interface.
+        """
+        self.epochs += 1
+        self.num_classes = max(
+            len(classes) if classes is not None else 0,
+            (int(np.max(y)) + 1), self.num_classes)
+        predictions = np.zeros((self.num_classes,))
+        max_weight = 0
+        weakest_expert_weight = 1
+        weakest_expert_index = None
+        
+        for i, exp in enumerate(self.experts):
+            yHat = exp.estimator.predict(X)
+            if np.any(yHat != y) and (self.epochs % self.period == 0):
+                exp.weight *= self.beta
+            
+            predictions[yHat] += exp.weight
+            max_weight = max(max_weight, exp.weight)
+
+            if exp.weight < weakest_expert_weight:
+                weakest_expert_index = i
+                weakest_expert_weight = exp.weight
+
+        yHat = np.array([np.argmax(predictions)])
+        if self.epochs % self.period == 0:
+            self._scale_weights(max_weight)
+            self._remove_experts()
+            if np.any(yHat != y):
+                if len(self.experts) == self.max_experts:
+                    self.experts.pop(weakest_expert_index)
+                self.experts.append(self._construct_new_expert())
+
+        # Train individual experts
+        for exp in self.experts:
+            exp.estimator.partial_fit(X, y, classes, weight)
+
+    def get_expert_predictions(self, X, classes=None, weight=None):
+        """
+        Returns predictions of each class for each expert.
+        In shape: (n_experts,)
+        """
+        return [exp.estimator.predict(X) for exp in self.experts]
+
+    def reset(self):
+        """
+        Reset this ensemble learner.
+        """
+        self.epochs = 0
```

Comment:
Some attributes are defined outside the `__init__` method:
`self.epochs`, `self.num_classes`, `self.experts`

Reply:
should I repeat the body of ```reset()``` in ```__init__``` ? (instead of calling reset in __init__)

Reply:
In my opinion, both are correct. Personally, I prefer calling `reset` from `__init__`.